### PR TITLE
Support various Python versions in yb_activate_virtualenv

### PR DESCRIPTION
**Pull request by mbautin**
- Allow passing a Python interpreter name to yb_activate_virtualenv
- Test Python virtual environment creation, activation, and deactivation, and module installation inside the virtual environments.

### CODE REVIEW

File: src/yugabyte-bash-common.sh [hunk 1]
```diff
@@ -367,22 +371,50 @@ expect_some_args() {
   fi
 }
 
-# Validates the number of arguments passed to its caller. Should also be passed all the caller's
+# This helps validate that the number of arguments passed to a function is within a certain range.
+# Should also be passed all the arguments.
 # arguments using "$@".
-# Example:
+# Examples:
+#   - To check is a function is passed exactly one argument:
```
Comment (jaki) [inline src/yugabyte-bash-common.sh]: style: s/check is a/check if a/

File: src/yugabyte-bash-common.sh [hunk 2]
```diff
@@ -367,22 +371,50 @@ expect_some_args() {
...
+#
 #   expect_num_args 1 "$@"
```
Comment (jaki) [inline src/yugabyte-bash-common.sh]: style: can this be indented to match the bulleted list level (2 spaces)?

File: src/yugabyte-bash-common.sh [hunk 3]
```diff
@@ -367,22 +371,50 @@ expect_some_args() {
...
+#
+#   - To check that the number of arguments passed to a function is within a certain range:
+#
+#   expect_num_args 1-2 "$@"
```
Comment (jaki) [inline src/yugabyte-bash-common.sh]: style: ditto

File: src/yugabyte-bash-common.sh [hunk 4]
```diff
@@ -367,22 +371,50 @@ expect_some_args() {
...
     yb_log_quiet=false
-    local error_msg="$calling_func_name expects $caller_expected_num_args arguments, got $#."
+    if [[ $num_args_lower_bound -eq $num_args_upper_bound ]]; then
+      local error_msg="$calling_func_name expects $expected_num_args arguments, got $#."
+    else
+      local error_msg=\
+"$calling_func_name expects from $num_args_lower_bound to $num_args_upper_bound arguments, got $#."
+    fi
     if [[ $# -eq 0 ]]; then
       error_msg+=" Check if \"\$@\" was included in the call to expect_num_args."
```
Comment (jaki) [inline src/yugabyte-bash-common.sh:416]: `$*` might be more appropriate here.

File: src/yugabyte-bash-common.sh [hunk 5]
```diff
@@ -694,21 +726,50 @@ yb_deactivate_virtualenv() {
...
+  fi
+  local venv_dir=$top_dir/$yb_virtualenv_basename
   yb_deactivate_virtualenv
   if [[ ! -d $venv_dir ]]; then
-    run_python -m pip install virtualenv --user
-    run_python -m virtualenv "$venv_dir"
+    if [[ $python_major_version -eq 3 ]]; then
+      "$python_interpreter" -m venv "$venv_dir"
+    else
+      "$python_interpreter" -m pip install virtualenv --user
```
Comment (jaki) [inline src/yugabyte-bash-common.sh:771]: Might want to log this?  I wouldn't want to have something installed without my knowledge.

File: test/test.sh [hunk 6]
```diff
@@ -98,8 +154,110 @@ yb_test_sha256sum() {
...
+  assert_incorrect_num_args 0 foo bar
+  assert_incorrect_num_args 1
+  assert_incorrect_num_args 1 foo bar
+  assert_incorrect_num_args 2
+  assert_incorrect_num_args 2 foo
+  assert_incorrect_num_args 2 foo bar baz
+}
+
+# Arguments:
+#   - Python interpreter (e.g. python.7 or python3)
```
Comment (jaki) [inline test/test.sh:175]: style: s/python.7/python2.7/

File: test/test.sh [hunk 7]
```diff
@@ -98,8 +154,110 @@ yb_test_sha256sum() {
...
+#   - Regular expression that that the Python version should match
```
Comment (jaki) [inline test/test.sh:176]: style: s/that that/that/

File: test/test.sh [hunk 8]
```diff
@@ -98,8 +154,110 @@ yb_test_sha256sum() {
...
+#   - One or more modules to install in the virtualenv.
+check_virtualenv() {
+  local python_interpreter=$1
+  local python_version_regex=$2
+  shift 2
+  local venv_parent_dir="$TEST_TMPDIR/${python_interpreter}_venv_parent_dir"
```
Comment (jaki) [inline test/test.sh]: style: quotes should be unnecessary here

File: test/test.sh [hunk 9]
```diff
@@ -98,8 +154,110 @@ yb_test_sha256sum() {
...
+  mkdir -p "$venv_parent_dir"
+  local requirement
+  for requirement in "$@"; do
+    echo "$requirement"
+  done >"$venv_parent_dir/requirements.txt"
+  local pip_list_output_path="$venv_parent_dir/pip_list_output.txt"
+  local python_interpreter_path_file="$venv_parent_dir/python_interpreter_path.txt"
```
Comment (jaki) [inline test/test.sh]: style: quotes should be unnecessary (this shows up several more times for the rest of the file)

File: test/test.sh [hunk 10]
```diff
@@ -98,8 +154,110 @@ yb_test_sha256sum() {
...
+  local python_interpreter_path_in_venv
+  python_interpreter_path_in_venv=$(<"$python_interpreter_path_file")
+  local actual_python_version
+  actual_python_version=$( "$python_interpreter_path_in_venv" --version 2>&1 )
+  assert_matches_regex "Python $python_version_regex" "$actual_python_version"
+}
+
+yb_test_activate_virtualenv() {
+  check_virtualenv python2.7 "2[.]7[.].*" psutil
+  check_virtualenv python3 "3[.].*" requests
```
Comment (jaki) [inline test/test.sh]: Improve regex ending `.*` to `\d+`

File: test/test.sh [hunk 11]
```diff
@@ -98,8 +154,110 @@ yb_test_sha256sum() {
...
+    command -v python >"$python_interpreter_path_file2"
+
+    yb_deactivate_virtualenv
+    pip list >"$pip_list_deactivated_output_path"
+  )
+
+  for requirement in "$@"; do
+    assert_egrep_no_results "^${requirement}[[:space:]]" "$pip_list_output_path1"
+    assert_egrep "^${requirement}[[:space:]]" "$pip_list_output_path2"
+    assert_egrep_no_results "^${requirement}[[:space:]]" "$pip_list_deactivated_output_path"
```
Comment (jaki) [inline test/test.sh:247]: Is this safe against non-virtual environments (esp. for requirement `requests`)?

File: test/test.sh [hunk 12]
```diff
@@ -124,10 +282,11 @@ trap cleanup EXIT
 
 global_exit_code=0
 test_fn_names=$(
-  declare -F | sed 's/^declare -f //g' | grep '^yb_test_'
+  declare -F | sed 's/^declare -f //g' | grep '^yb_test_' | sort
```
Comment (jaki) [inline test/test.sh]: Not sure why `/g` is needed here because `/^` guarantees one match per line.

**Review by jaki — APPROVED** [review]
Tests pass; looks good; left some comments.

File: src/yugabyte-bash-common.sh (hunk 4 above)
Comment (mbautin) [reply src/yugabyte-bash-common.sh:416]: No, `"$@"` is exactly what the caller needs to specify.

**Pull request merged**